### PR TITLE
GEODE-4327: Added mod and arithmetic functionality to oql engine

### PR DESCRIPTION
**Pull request by jhuynh1**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledDivision.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.geode.cache.query.internal.parse.OQLLexerTokenTypes;
+
+public class CompiledDivision extends CompiledArithmetic implements OQLLexerTokenTypes {
+
+  CompiledDivision(CompiledValue left, CompiledValue right, int op) {
+    super(left, right, op);
+  }
+
+  public int getType() {
+    return MULTIPLICATION;
```
Comment (nabarunnag) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledDivision.java]: I think this should be DIVISION

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledDivision.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  @Override
+  public Double evaluateArithmeticOperation(Double n1, Double n2) {
+    return n1 / n2;
+  }
+
+  @Override
+  public Float evaluateArithmeticOperation(Float n1, Float n2) {
+    return n1 / n2;
```
Comment (nabarunnag) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledDivision.java]: Just of curiosity, do we need to handle division by zero exception somewhere ? Or things like overflowing the int limit after an arithmetic operation.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+      SelectResults results = (SelectResults) queryService
+          .newQuery("select * from /portfolio p where p.ID / null = 2").execute();
+      fail();
+    } catch (QueryInvocationTargetException qite) {
+      assertEquals("Cannot evaluate arithmetic operations on null values", qite.getMessage());
+    }
+  }
+
+  @Test
+  public void divideOperationShouldCorrectlyExecuteInQuery() throws Exception {
```
Comment (nabarunnag) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java]: tests for division by zero needed ?

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    SelectResults results = (SelectResults) queryService
+        .newQuery("select * from /portfolio p where p.ID + 1 = 2").execute();
+    assertEquals(1, results.size());
+  }
+
+  @Test
+  public void addOperationShouldThrowExceptionWhenAddingNonNumeric() throws Exception {
+    try {
+      SelectResults results = (SelectResults) queryService
+          .newQuery("select * from /portfolio p where p.ID + '1' = 2").execute();
```
Comment (nabarunnag) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java]: tests with addition causing to overflow the int limit ?

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java (hunk 3 above)
Comment (jhuynh1) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java]: Sure I can add a test for that

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java (hunk 4 above)
Comment (jhuynh1) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java]: I think we will just use whatever java does with overflown values.  If we want to upcast after addition or multiplication then that can be a different feature?

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledDivision.java (hunk 1 above)
Comment (jhuynh1) [reply geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledDivision.java]: Ok, no need to yell ;-)

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java (hunk 3 above)
Comment (metatype) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java]: What about tests for operator precedence?

`select * from /portfolio p where p.ID + 2 / 3 = 1`

Comment (jhuynh1) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/ArithmeticOperationsIntegrationTest.java]: Added some operator precedence tests

**Pull request merged**